src/Microsoft.AspNetCore.SignalR.Core/HubGroupList.cs

```diff
@@ -0,0 +1,77 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.AspNetCore.SignalR
+{
+    public class HubGroupList : IReadOnlyCollection<ConcurrentDictionary<string, HubConnectionContext>>
+    {
+        private readonly ConcurrentDictionary<string, ConcurrentDictionary<string, HubConnectionContext>> _groups = new ConcurrentDictionary<string, ConcurrentDictionary<string, HubConnectionContext>>();
+
+        public ConcurrentDictionary<string, HubConnectionContext> this[string groupName]
+        {
+            get
+            {
+                if (_groups.TryGetValue(groupName, out var group))
+                {
+                    return group;
+                }
+                return null;
+            }
+        }
+
+        public void Add(HubConnectionContext connection, string groupName)
+        {
+            CreateOrUpdateGroupWithConnection(groupName, connection);
+        }
+
+        public void Remove(string connectionId, string groupName)
+        {
+            if (_groups.TryGetValue(groupName, out var connections))
+            {
+                if (connections.TryRemove(connectionId, out var _) && connections.Count == 0)
+                {
+                    if (_groups.TryRemove(groupName, out var newlyAddedConnections) && newlyAddedConnections != null)
```

Comment:
This doesn't look great. Possibly recreating the group you just deleted... can you think of a better way to do this?

Reply:
Let me explain.
When we remove empty group, at same time possible someone joined group and we lost that connection, but by this code we get newly added connections and recreate group.

Reply:
I know why you do it, I'm just saying it looks ugly and if there is a better solution we should do it.

Reply:
OK, I will think about better solution

Reply:
Don't put too much time into it though :)
I also am hitting a similar issue in something I'm doing, so I'll try to think of a better solution too.

Reply:
I did other comments, will commit

Reply:
Done

Reply:
@BrennanConroy what do you think about this?
`((ICollection<KeyValuePair<string, ConcurrentDictionary<string, HubConnectionContext>>>)_groups).Remove(new KeyValuePair<string, ConcurrentDictionary<string, HubConnectionContext>>(groupName, new ConcurrentDictionary<string, HubConnectionContext>()));`

Source: https://blogs.msdn.microsoft.com/pfxteam/2011/04/02/little-known-gems-atomic-conditional-removals-from-concurrentdictionary/

Reply:
That looks promising!

Reply:
Yes, I think so.
I will change to this.

After this update what prevent to merge this PR?

Reply:
I want to make users by same way, after making users with this way we will have full indexed clients selection

Reply:
One more thing.
Because ```ICollection.Remove()``` method compare key and value, "promising" way will now work in this case, because we have empty dictionary but with another HashCode. So this will not work.